File: shared/src/main/java/de/dreier/mytargets/shared/models/TrainingInfo.java
/*
 * Copyright (C) 2017 Florian Dreier
 *
 * This file is part of MyTargets.
 *
 * MyTargets is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2
 * as published by the Free Software Foundation.
 *
 * MyTargets is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 */
package de.dreier.mytargets.shared.models;

import android.content.Context;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import java.util.List;

import de.dreier.mytargets.shared.R;
import de.dreier.mytargets.shared.models.augmented.AugmentedEnd;
import de.dreier.mytargets.shared.models.augmented.AugmentedRound;
import de.dreier.mytargets.shared.models.augmented.AugmentedTraining;
import de.dreier.mytargets.shared.models.db.Round;

public class TrainingInfo {
    @Nullable
    public String title;
    public int roundCount;
    public AugmentedRound round;

    public TrainingInfo() {
    }

    public TrainingInfo(@NonNull AugmentedTraining training, AugmentedRound round) {
        this.round = round;
        this.title = training.getTraining().getTitle();
        this.roundCount = training.getRounds().size();
    }

    @NonNull
    public String getRoundDetails(@NonNull Context context) {
        if (round.getEnds().isEmpty()) {
            return context.getResources()
                    .getQuantityString(R.plurals.rounds, roundCount, roundCount);
        } else {
            return round.getReachedScore().toString();
        }
    }

    @NonNull
    public String getEndDetails(@NonNull Context context) {
        Round simpleRound = round.getRound();
        List<AugmentedEnd> ends = round.getEnds();
        if (ends.isEmpty()) {
            if (simpleRound.getMaxEndCount() == null) {
                return context.getResources()
                        .getQuantityString(R.plurals.arrows_per_end, simpleRound.getShotsPerEnd(), simpleRound
                                .getShotsPerEnd());
            } else {
                return context.getResources()
                        .getQuantityString(R.plurals.ends_arrow, simpleRound.getShotsPerEnd(), simpleRound
                                .getMaxEndCount(), simpleRound.getShotsPerEnd());
            }
        } else {
            return context.getResources()
                    .getQuantityString(R.plurals.ends_arrow, simpleRound.getShotsPerEnd(), ends
                            .size(), simpleRound.getShotsPerEnd());
        }
    }
}

File: shared/src/main/java/de/dreier/mytargets/shared/models/TrainingInfo.kt
/*
 * Copyright (C) 2018 Florian Dreier
 *
 * This file is part of MyTargets.
 *
 * MyTargets is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2
 * as published by the Free Software Foundation.
 *
 * MyTargets is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 */
package de.dreier.mytargets.shared.models

import android.content.Context
import android.os.Parcel
import android.os.Parcelable
import de.dreier.mytargets.shared.R
import de.dreier.mytargets.shared.models.augmented.AugmentedRound
import de.dreier.mytargets.shared.models.augmented.AugmentedTraining

data class TrainingInfo(
        var title: String? = null,
        var roundCount: Int = 0,
        var round: AugmentedRound) : Parcelable {

    constructor(training: AugmentedTraining, round: AugmentedRound) : this(
            title = training.training.title,
            roundCount = training.rounds.size,
            round = round
    )

    constructor(source: Parcel) : this(
            source.readString(),
            source.readInt(),
            source.readParcelable<AugmentedRound>(AugmentedRound::class.java.classLoader)!!
    )

    fun getRoundDetails(context: Context): String {
        return if (round.ends.isEmpty()) {
            context.resources
                    .getQuantityString(R.plurals.rounds, roundCount, roundCount)
        } else {
            round.round.score.toString()
        }
    }

    fun getEndDetails(context: Context): String {
        val (_, _, _, shotsPerEnd, maxEndCount) = round.round
        val ends = round.ends
        return if (ends.isEmpty()) {
            if (maxEndCount == null) {
                context.resources
                        .getQuantityString(R.plurals.arrows_per_end, shotsPerEnd, shotsPerEnd)
            } else {
                context.resources
                        .getQuantityString(R.plurals.ends_arrow, shotsPerEnd, maxEndCount, shotsPerEnd)
            }
        } else {
            context.resources
                    .getQuantityString(R.plurals.ends_arrow, shotsPerEnd, ends
                            .size, shotsPerEnd)
        }
    }

    override fun describeContents() = 0

    override fun writeToParcel(dest: Parcel, flags: Int) = with(dest) {
        writeString(title)
        writeInt(roundCount)
        writeParcelable(round, 0)
    }

    companion object {
        @JvmField
        val CREATOR: Parcelable.Creator<TrainingInfo> = object : Parcelable.Creator<TrainingInfo> {
            override fun createFromParcel(source: Parcel): TrainingInfo = TrainingInfo(source)
            override fun newArray(size: Int): Array<TrainingInfo?> = arrayOfNulls(size)
        }
    }
}
